### PR TITLE
[minor] SCA with Php Inspections (EA Ultimate)

### PR DESCRIPTION
**Pull request by kalessil**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | no
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | n/a

Addressed some of CS and performance related findings.
Please reject what's looks risky.

### CODE REVIEW

**Review by SilverFire — CHANGES_REQUESTED** [review]

File: framework/base/Component.php [hunk 1]
```diff
@@ -602,12 +602,13 @@ public function trigger($name, Event $event = null)
     {
         $this->ensureBehaviors();
 
-        $eventHandlers = [];
+        $eventHandlers = [[]];
         foreach ($this->_eventWildcards as $wildcard => $handlers) {
             if (StringHelper::matchWildcard($wildcard, $name)) {
-                $eventHandlers = array_merge($eventHandlers, $handlers);
+                $eventHandlers[] = $handlers;
```
[the remark this answers is not in the document]
Comment (kalessil) [reply framework/base/Component.php]: Ok, will revert

File: framework/console/controllers/AssetController.php [hunk 2]
```diff
@@ -646,7 +646,7 @@ protected function adjustCssUrl($cssContent, $inputFilePath, $outputFilePath)
             $fullMatch = $matches[0];
             $inputUrl = $matches[1];
 
-            if (strpos($inputUrl, '/') === 0 || strpos($inputUrl, '#') === 0 || preg_match('/^https?:\/\//i', $inputUrl) || preg_match('/^data:/i', $inputUrl)) {
+            if (strncmp($inputUrl, '/', 1) === 0 || strncmp($inputUrl, '#', 1) === 0 || preg_match('/^https?:\/\//i', $inputUrl) || preg_match('/^data:/i', $inputUrl)) {
```
[the remark this answers is not in the document]
Comment (kalessil) [reply framework/console/controllers/AssetController.php]: Yes, Yii, in general, uses this and not much strpos

File: framework/base/Component.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (kalessil) [reply framework/base/Component.php]: Done

**Comment by kalessil** [comment]
Build now fails with HHVM only, doesn't seem to be due to my changes:

```
1) yiiunit\framework\web\UserTest::testLoginRequired
ps_files_cleanup_dir: opendir(/var/lib/hhvm/sessions) failed: Permission denied (13)
```

**Review by samdark — APPROVED** [review]

File: framework/web/Request.php [hunk 3]
```diff
@@ -809,7 +809,7 @@ public function getScriptUrl()
             } elseif (isset($_SERVER['PHP_SELF']) && ($pos = strpos($_SERVER['PHP_SELF'], '/' . $scriptName)) !== false) {
                 $this->_scriptUrl = substr($_SERVER['SCRIPT_NAME'], 0, $pos) . '/' . $scriptName;
             } elseif (!empty($_SERVER['DOCUMENT_ROOT']) && strpos($scriptFile, $_SERVER['DOCUMENT_ROOT']) === 0) {
-                $this->_scriptUrl = str_replace('\\', '/', str_replace($_SERVER['DOCUMENT_ROOT'], '', $scriptFile));
+                $this->_scriptUrl = str_replace(array($_SERVER['DOCUMENT_ROOT'], '\\'), array('', '/'), $scriptFile);
```
Comment (schmunk42) [inline framework/web/Request.php]: How about using short array syntax here `[]`?

File: framework/web/Response.php [hunk 4]
```diff
@@ -767,7 +766,7 @@ public function xSendFile($filePath, $attachmentName = null, $options = [])
      */
     protected function getDispositionHeaderValue($disposition, $attachmentName)
     {
-        $fallbackName = str_replace('"', '\\"', str_replace(['%', '/', '\\'], '_', Inflector::transliterate($attachmentName, Inflector::TRANSLITERATE_LOOSE)));
+        $fallbackName = str_replace(array('%', '/', '\\', '"'), array('_', '_', '_', '\\"'), Inflector::transliterate($attachmentName, Inflector::TRANSLITERATE_LOOSE));
```
Comment (schmunk42) [inline framework/web/Response.php]: For more complex replacements, how about splitting it into three lines?

```
$fallbackName = str_replace(
    ['%', '/', '\\', '"'], 
    ['_', '_', '_', '\\"'], 
    Inflector::transliterate($attachmentName, Inflector::TRANSLITERATE_LOOSE)
);
```

**Comment by SilverFire** [comment]
> Build now fails with HHVM only, doesn't seem to be due to my changes

No, it's not your fail

**Review by dmirogin — APPROVED** [review]

File: framework/web/Request.php (hunk 3 above)
Comment (kalessil) [reply framework/web/Request.php]: Makes sense, will update.

File: framework/web/Response.php (hunk 4 above)
Comment (kalessil) [reply framework/web/Response.php]: Good idea =)

**Review by schmunk42 — APPROVED** [review]

**Review by SilverFire — APPROVED** [review]

**Comment by samdark** [comment]
Thank you!

File: framework/db/ActiveQuery.php [hunk 5]
```diff
@@ -762,7 +762,7 @@ public function orOnCondition($condition, $params = [])
      */
     public function viaTable($tableName, $link, callable $callable = null)
     {
-        $modelClass = $this->primaryModel !== null ? get_class($this->primaryModel) : get_class();
+        $modelClass = $this->primaryModel !== null ? get_class($this->primaryModel) : __CLASS__;
```
Comment (cebe) [inline framework/db/ActiveQuery.php]: This looks really weird, using ActiveQuery and modelclass does not make any sense here. If primaryModel is null, it should fail, one can not call viaTable on a query that is not a relation...

Comment (cebe) [reply framework/db/ActiveQuery.php]: found the original commit that added this behavior here: https://github.com/yiisoft/yii2/commit/64d8af61a64db0b00d498800e1f20e62b4473329#diff-1776bc9373c1dbb0d6e77ba4d8081a81L762

Comment (cebe) [reply framework/db/ActiveQuery.php]: not related to this PR; created an issue: https://github.com/yiisoft/yii2/issues/15876

**Pull request merged**